### PR TITLE
fix: correctly poll `BasicBlockDownloader`

### PR DESCRIPTION
**Pull request by klkvr**
Right now if we schedule block download, on next poll of `BasicBlockDownloader` we'll return `Poll:Ready` with `NewDownloadStarted` event and never register waker with actual downloading future.

Found this to be an issue when writing e2e tests, not sure if can occur during normal node operation

### CODE REVIEW

**Comment by mattsse** [comment]
> we'll return Poll:Ready with NewDownloadStarted event and never register waker with actual downloading future.

hmm, this is expected because a ready indicates that there can be more work and the caller is supposed to call it again

which I think we do:

https://github.com/paradigmxyz/reth/blob/5532d34ec972ed734267616ef629e57d0ddce65b/crates/engine/tree/src/engine.rs#L115-L120

> Found this to be an issue when writing e2e tests, not sure if can occur during normal node operation
which issue did you encounter?

**Comment by klkvr** [comment]
ah, then I think the issue is in this line:
https://github.com/paradigmxyz/reth/blob/5532d34ec972ed734267616ef629e57d0ddce65b/crates/engine/tree/src/engine.rs#L116

if we get `Poll::Ready` with any other `DownloadOutcome` variant we never poll it again

**Comment by klkvr** [comment]
the issue I've seen is that block fetching future was not being polled

**Comment by klkvr** [comment]
Updated logic to instead make sure that we drain downloader events in `EngineHandler`

File: crates/engine/tree/src/engine.rs [hunk 1]
```diff
@@ -113,9 +113,11 @@ where
             }
 
             // advance the downloader
-            if let Poll::Ready(DownloadOutcome::Blocks(blocks)) = self.downloader.poll(cx) {
-                // delegate the downloaded blocks to the handler
-                self.handler.on_event(FromEngine::DownloadedBlocks(blocks));
+            if let Poll::Ready(outcome) = self.downloader.poll(cx) {
+                if let DownloadOutcome::Blocks(blocks) = outcome {
```
Comment (mattsse) [inline crates/engine/tree/src/engine.rs:117]: ah, yes of course lol

initial I wanted to bubble up the newdownload started event, but came to the conclusion that this isn't very useful

**Review by mattsse — APPROVED** [review]

**Pull request merged**